lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -507,55 +510,94 @@ TEE_Result crypto_hash_final(void *ctx, uint32_t algo, uint8_t *digest,
  ******************************************************************************/
 
 #if defined(_CFG_CRYPTO_WITH_ACIPHER)
-struct bignum *crypto_bignum_allocate(size_t size_bits __unused)
+#define ciL		(sizeof(mbedtls_mpi_uint))	/* chars in limb  */
+#define biL		(ciL << 3)			/* bits  in limb  */
+#define biH		(ciL << 2)			/* half limb size */
+#define BITS_TO_LIMBS(i)		((i) / biL + ((i) % biL != 0))
+
+#define LMD_MAX_BITS_PER_VARIABLE	(4096)
+
+size_t crypto_bignum_num_bytes(struct bignum *a)
 {
-	return NULL;
+	assert(a != NULL);
+	return mbedtls_mpi_size((const mbedtls_mpi *)a);
 }
 
-TEE_Result crypto_bignum_bin2bn(const uint8_t *from __unused,
-				size_t fromsize __unused,
-				struct bignum *to __unused)
+size_t crypto_bignum_num_bits(struct bignum *a)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	assert(a != NULL);
+	return mbedtls_mpi_bitlen((const mbedtls_mpi *)a);
 }
 
-size_t crypto_bignum_num_bytes(struct bignum *a __unused)
+int32_t crypto_bignum_compare(struct bignum *a, struct bignum *b)
 {
-	return 0;
+	int ret;
+
+	assert(a != NULL);
+	assert(b != NULL);
+	ret = mbedtls_mpi_cmp_mpi((const mbedtls_mpi *)a,
+		(const mbedtls_mpi *)b);
+	if (ret < 0)
+		return -1;
+	else if (ret > 0)
+		return 1;
+	else
+		return 0;
 }
 
-size_t crypto_bignum_num_bits(struct bignum *a __unused)
+void crypto_bignum_bn2bin(const struct bignum *from, uint8_t *to)
 {
-	return 0;
+	size_t len;
+
+	assert(from != NULL);
+	assert(to != NULL);
+	len = crypto_bignum_num_bytes((struct bignum *)from);
+	mbedtls_mpi_write_binary((mbedtls_mpi *)from, to, len);
 }
 
-void crypto_bignum_bn2bin(const struct bignum *from __unused,
-			  uint8_t *to __unused)
+TEE_Result crypto_bignum_bin2bn(const uint8_t *from, size_t fromsize,
+			 struct bignum *to)
 {
+	assert(from != NULL);
+	assert(to != NULL);
+	if (mbedtls_mpi_read_binary((mbedtls_mpi *)to, from, fromsize) != 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+	return TEE_SUCCESS;
 }
 
-void crypto_bignum_copy(struct bignum *to __unused,
-			const struct bignum *from __unused)
+void crypto_bignum_copy(struct bignum *to, const struct bignum *from)
 {
+	assert(from != NULL);
+	assert(to != NULL);
+	mbedtls_mpi_copy((mbedtls_mpi *)to, (const mbedtls_mpi *)from);
 }
 
-void crypto_bignum_free(struct bignum *a)
+struct bignum *crypto_bignum_allocate(size_t size_bits)
 {
-	if (a)
-		panic();
+	mbedtls_mpi *bn;
+
+	bn = calloc(1, sizeof(mbedtls_mpi));
+	if (bn == NULL)
+		return NULL;
+	mbedtls_mpi_init(bn);
+	if (mbedtls_mpi_grow(bn, BITS_TO_LIMBS(size_bits)) != 0)
+		return NULL;
```

Comment:
`bn` leakage

Reply:
`bn` is just allocated with `calloc()`, if we return NULL here without freeing `bn` we have a memory leak.

Reply:
Yes, I will add `free(bn)` here.